File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/valid-sudoku/code/python/valid_sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # O(1) time — Track digits per row, col, 3x3 box
        seen = set()
        for i in range(9):
            for j in range(9):
                d = board[i][j]
                if d == '.': continue
                row = (d, 'row', i)
                col = (d, 'col', j)
                box = (d, 'box', i//3, j//3)
                if row in seen or col in seen or box in seen:
                    return False
                seen.update([row, col, box])
        return True

File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/valid-sudoku/code/java/ValidSudoku.java
class Solution {
    // O(1) time (81 cells) — Track seen digits per row, col, box
    public boolean isValidSudoku(char[][] board) {
        Set<String> seen = new HashSet<>();
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;
                char d = board[i][j];
                if (!seen.add(d + " row " + i) ||
                    !seen.add(d + " col " + j) ||
                    !seen.add(d + " box " + i/3 + "-" + j/3))
                    return false;
            }
        }
        return true;
    }
}